python/pyarrow/table.pxi

```diff
@@ -2983,6 +2985,100 @@ cdef class RecordBatch(_Tabular):
                     <ArrowArray*> c_ptr, c_schema))
         return pyarrow_wrap_batch(c_batch)
 
+    def __arrow_c_array__(self, requested_schema=None):
+        """
+        Get a pair of PyCapsules containing a C ArrowArray representation of the object.
+
+        Parameters
+        ----------
+        requested_schema : PyCapsule | None
+            A PyCapsule containing a C ArrowSchema representation of a requested
+            schema. PyArrow will attempt to cast the batch to this schema.
+            If None, the schema will be returned as-is, with a schema matching the
+            one returned by :meth:`__arrow_c_schema__()`.
+
+        Returns
+        -------
+        Tuple[PyCapsule, PyCapsule]
+            A pair of PyCapsules containing a C ArrowSchema and ArrowArray,
+            respectively.
+        """
+        cdef:
+            ArrowArray* c_array
+            ArrowSchema* c_schema
+
+        if requested_schema is not None:
+            target_schema = Schema._import_from_c_capsule(requested_schema)
+
+            if target_schema != self.schema:
+                try:
+                    # We don't expose .cast() on RecordBatch, only on Table.
+                    casted_batch = Table.from_batches([self]).cast(
+                        target_schema, safe=True).to_batches()[0]
+                    inner_batch = pyarrow_unwrap_batch(casted_batch)
+                except ArrowInvalid as e:
+                    raise ValueError(
+                        f"Could not cast {self.schema} to requested schema {target_schema}"
```

Comment:
```suggestion
                        f"Could not cast {self.schema} to requested schema {target_schema}: {e}"
```

Reply:
Does Cython not do [exception chaining](https://docs.python.org/3/tutorial/errors.html#exception-chaining)?

Reply:
I don't know, but it doesn't help if one merely prints the exception message. Here, we are really re-raising the original error with a bit more context in the exception message, so IMHO it makes sense to keep the original message as well.

Reply:
My suggestion was mostly for improving clarity (I personally think it will be easier to understand if the final error message has the information about why the cast failed, than having to look somewhere inside the traceback, which can be quite long in IPython, in contrast to the concise example in the python docs), and for consistency, as the same is done for the Array version.